File: Languages/Swift/hashgen.swift
import Foundation
import CommonCrypto

// Hello, World!
print("Hello, World!")

struct FileHashResult: Codable {
    let filename: String
    let filepath: String
    let size: Int64
    let hash: String
    let status: String
    let error: String?
    
    enum CodingKeys: String, CodingKey {
        case filename, filepath, size, status, error
        case hash
    }
}

struct HashResults: Codable {
    let totalFiles: Int
    let successfulHashes: Int
    let failedHashes: Int
    let algorithm: String
    let processingTimeSeconds: Double
    let averageTimePerFileMs: Double
    let timestamp: String
    let directory: String
    let results: [FileHashResult]
    
    enum CodingKeys: String, CodingKey {
        case totalFiles, successfulHashes, failedHashes, results
        case algorithm, timestamp, directory
        case processingTimeSeconds, averageTimePerFileMs
    }
}

func calculateFileHash(filepath: String) -> String? {
    guard let data = try? Data(contentsOf: URL(fileURLWithPath: filepath)) else {
        return nil
    }
    
    var hash = [UInt8](repeating: 0, count: Int(CC_SHA256_DIGEST_LENGTH))
    
    data.withUnsafeBytes { bytes in
        guard let baseAddress = bytes.baseAddress else { return }
        CC_SHA256(baseAddress, CC_LONG(data.count), &hash)
    }
    
    return hash.map { String(format: "%02x", $0) }.joined()
}

func getFileSize(filepath: String) -> Int64 {
    do {
        let attributes = try FileManager.default.attributesOfItem(atPath: filepath)
        return attributes[.size] as? Int64 ?? 0
    } catch {
        return 0
    }
}

func getTxtFiles(directory: String) -> [String] {
    let fileManager = FileManager.default
    
    guard fileManager.fileExists(atPath: directory) else {
        print("Directory not found: \(directory)")
        return []
    }
    
    do {
        let files = try fileManager.contentsOfDirectory(atPath: directory)
            .filter { $0.hasSuffix(".txt") }
            .map { "\(directory)/\($0)" }
            .sorted()
        return files
    } catch {
        print("Error reading directory: \(error)")
        return []
    }
}

func getTimestamp() -> String {
    let formatter = DateFormatter()
    formatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
    return formatter.string(from: Date())
}

func hashFilesInDirectory(directory: String) {
    let files = getTxtFiles(directory: directory)
    
    if files.isEmpty {
        print("No .txt files found in directory: \(directory)")
        return
    }
    
    print("Found \(files.count) files to hash")
    
    var results = [FileHashResult]()
    var successfulHashes = 0
    var failedHashes = 0
    
    let startTime = Date()
    
    for (index, filepath) in files.enumerated() {
        let filename = URL(fileURLWithPath: filepath).lastPathComponent
        let size = getFileSize(filepath: filepath)
        
        var result: FileHashResult
        
        if let hash = calculateFileHash(filepath: filepath) {
            result = FileHashResult(
                filename: filename,
                filepath: filepath,
                size: size,
                hash: hash,
                status: "SUCCESS",
                error: nil
            )
            successfulHashes += 1
        } else {
            result = FileHashResult(
                filename: filename,
                filepath: filepath,
                size: size,
                hash: "",
                status: "FAILED",
                error: "Hash calculation failed"
            )
            failedHashes += 1
        }
        
        results.append(result)
        
        // Progress indicator
        if (index + 1) % 100 == 0 {
            print("Processed \(index + 1)/\(files.count) files...")
        }
    }
    
    let endTime = Date()
    let processingTime = endTime.timeIntervalSince(startTime)
    let averageTimePerFile = (processingTime / Double(files.count)) * 1000
    
    let finalResults = HashResults(
        totalFiles: files.count,
        successfulHashes: successfulHashes,
        failedHashes: failedHashes,
        algorithm: "sha256",
        processingTimeSeconds: processingTime,
        averageTimePerFileMs: averageTimePerFile,
        timestamp: getTimestamp(),
        directory: directory,
        results: results
    )
    
    // Save to file
    let encoder = JSONEncoder()
    encoder.outputFormatting = .prettyPrinted
    
    do {
        let data = try encoder.encode(finalResults)
        try data.write(to: URL(fileURLWithPath: "hash_results.json"))
        print("Results saved to: hash_results.json")
    } catch {
        print("Error saving results: \(error)")
    }
    
    print("\nHashing completed!")
    print("Total files: \(finalResults.totalFiles)")
    print("Successful: \(finalResults.successfulHashes)")
    print("Failed: \(finalResults.failedHashes)")
    print(String(format: "Processing time: %.3f seconds", finalResults.processingTimeSeconds))
    print(String(format: "Average time per file: %.2f ms", finalResults.averageTimePerFileMs))
}

hashFilesInDirectory(directory: "../hashfiles")

File: Languages/Haskell/HashGen.hs
module Main where

import System.IO
import System.Directory
import System.FilePath
import Data.Digest.Pure.SHA
import qualified Data.ByteString.Lazy as BL
import qualified Data.ByteString.Char8 as BC
import Data.Time
import Control.Exception
import Control.Monad
import Data.List (sort)
import Data.Aeson
import Data.Aeson.Types

-- Hello, World!
hello :: IO ()
hello = putStrLn "Hello, World!"

data FileHashResult = FileHashResult
    { filename :: String
    , filepath :: String
    , size :: Integer
    , hash :: String
    , status :: String
    , error :: Maybe String
    } deriving (Show, Eq)

instance ToJSON FileHashResult where
    toJSON (FileHashResult filename' filepath' size' hash' status' error') =
        object [ "filename" .= filename'
               , "filepath" .= filepath'
               , "size" .= size'
               , "algorithm" .= ("sha256" :: String)
               , "hash" .= hash'
               , "status" .= status'
               , "error" .= error'
               ]

data HashResults = HashResults
    { totalFiles :: Int
    , successfulHashes :: Int
    , failedHashes :: Int
    , algorithm :: String
    , processingTimeSeconds :: Double
    , averageTimePerFileMs :: Double
    , timestamp :: String
    , directory :: String
    , results :: [FileHashResult]
    } deriving (Show, Eq)

instance ToJSON HashResults where
    toJSON (HashResults totalFiles' successfulHashes' failedHashes' algorithm' 
                processingTimeSeconds' averageTimePerFileMs' timestamp' directory' results') =
        object [ "total_files" .= totalFiles'
               , "successful_hashes" .= successfulHashes'
               , "failed_hashes" .= failedHashes'
               , "algorithm" .= algorithm'
               , "processing_time_seconds" .= processingTimeSeconds'
               , "average_time_per_file_ms" .= averageTimePerFileMs'
               , "timestamp" .= timestamp'
               , "directory" .= directory'
               , "results" .= results'
               ]

calculateFileHash :: FilePath -> IO String
calculateFileHash filepath = do
    content <- BL.readFile filepath
    return $ show $ sha256 content

getFileSize :: FilePath -> IO Integer
getFileSize filepath = do
    exists <- doesFileExist filepath
    if exists
        then getFileSize filepath
        else return 0

getTxtFiles :: FilePath -> IO [FilePath]
getTxtFiles directory = do
    exists <- doesDirectoryExist directory
    if not exists
        then do
            putStrLn $ "Directory not found: " ++ directory
            return []
        else do
            allFiles <- listDirectory directory
            let txtFiles = map (directory </>) $ filter (".txt" `isSuffixOf`) allFiles
            return $ sort txtFiles

getTimestamp :: IO String
getTimestamp = do
    currentTime <- getCurrentTime
    return $ formatTime defaultTimeLocale "%Y-%m-%d %H:%M:%S" currentTime

hashFilesInDirectory :: FilePath -> IO ()
hashFilesInDirectory directory = do
    files <- getTxtFiles directory
    
    if null files
        then putStrLn $ "No .txt files found in directory: " ++ directory
        else do
            putStrLn $ "Found " ++ show (length files) ++ " files to hash"
            
            startTime <- getCurrentTime
            
            results <- mapM processFile files
            
            endTime <- getCurrentTime
            let processingTime = realToFrac $ diffUTCTime endTime startTime :: Double
            let averageTimePerFile = (processingTime / fromIntegral (length files)) * 1000
            
            timestamp <- getTimestamp
            
            let successful = length $ filter (\r -> status r == "SUCCESS") results
            let failed = length - successful
            
            let finalResults = HashResults
                    { totalFiles = length files
                    , successfulHashes = successful
                    , failedHashes = failed
                    , algorithm = "sha256"
                    , processingTimeSeconds = processingTime
                    , averageTimePerFileMs = averageTimePerFile
                    , timestamp = timestamp
                    , directory = directory
                    , results = results
                    }
            
            -- Save to file
            let outputFile = "hash_results.json"
            BL.writeFile outputFile (encode finalResults)
            putStrLn $ "Results saved to: " ++ outputFile
            
            putStrLn "\nHashing completed!"
            putStrLn $ "Total files: " ++ show (length files)
            putStrLn $ "Successful: " ++ show successful
            putStrLn $ "Failed: " ++ show failed
            putStrLn $ "Processing time: " ++ show (round (processingTime * 1000) / 1000) ++ " seconds"
            putStrLn $ "Average time per file: " ++ show (round (averageTimePerFile * 100) / 100) ++ " ms"
            
            -- Progress indicator
            mapM_ (\i -> when (i `mod` 100 == 0) $ 
                     putStrLn $ "Processed " ++ show i ++ "/" ++ show (length files) ++ " files...") [1..length files]

processFile :: FilePath -> IO FileHashResult
processFile filepath = do
    let filename = takeFileName filepath
    size <- getFileSize filepath
    
    result <- catch (calculateFileHash filepath) (\(_ :: SomeException) -> return "")
    
    if not (null result)
        then return $ FileHashResult filename filepath size result "SUCCESS" Nothing
        else return $ FileHashResult filename filepath size "" "FAILED" (Just "Hash calculation failed")

main :: IO ()
main = do
    hello
    hashFilesInDirectory "../hashfiles"
